### PR TITLE
lwip_sock: provide implementation for `sock_*_recv_buf()`

### PR DESCRIPTION
**Pull request by miri64**
### Contribution description
This provides an implementation for lwIP of the zero-copy receive functions (see #13616). As the implementation of sock_*_recv() now uses those functions, I only adapted the tests minimally to test the success case to acknowledge the function is there.

I am not 100% sure it is okay, not to loop over the netbuf in `sock_udp` (for `sock_ip` I did not), but testing will tell.

### Testing procedure
Should still pass (**only `LWIP_IPV6=1` is tested by the CI**):
```sh
BOARD="<pick one>"
for test in tests/lwip_sock_{ip,udp}; do
    LWIP_IPV6=1 make -C ${test} clean flash -j16 test;
    LWIP_IPV4=1 make -C ${test} clean flash -j16 test;
    LWIP_IPV4=1 LWIP_IPV6=1 make -C ${test} clean flash -j16 test;
done
```

Also **please, please, please** test with (IP) fragmented packets (so larger than MTU) for `sock_udp` with `tests/lwip`. I did not have any time to test that right now, but we really should, as I removed a copying loop and replaced it just with a pointer assignment (see https://github.com/RIOT-OS/RIOT/pull/13701#issuecomment-630860982 for procedure).

### Issues/PRs references
Follow-up to #13616.
Requires ~~#13779~~ (merged).

### CODE REVIEW

File: pkg/lwip/contrib/sock/udp/lwip_sock_udp.c [hunk 1]
```diff
@@ -113,12 +130,10 @@ ssize_t sock_udp_recv(sock_udp_t *sock, void *data, size_t max_len,
...
         remote->port = buf->port;
     }
-    /* copy data */
-    for (struct pbuf *q = buf->p; q != NULL; q = q->next) {
-        memcpy(data_ptr, q->payload, q->len);
-        data_ptr += q->len;
-    }
-    netbuf_delete(buf);
+    assert(buf->p->next == NULL);
+    assert(buf->p->len == res);
```
Comment (miri64) [inline pkg/lwip/contrib/sock/udp/lwip_sock_udp.c]: Sadly, this doesn't work :-/ as soon as the received UDP payload is larger then 530 with IPv6, this assertions are hit :-(. I'll try to find out if there is another way out.

Comment (benpicco) [reply pkg/lwip/contrib/sock/udp/lwip_sock_udp.c]: Is this still the case?

Comment (miri64) [reply pkg/lwip/contrib/sock/udp/lwip_sock_udp.c]: Nope. This was what #13779 and the latest fixups were about.

**Comment by benpicco** [comment]
If this can be reviewed now, just squash - there hasn't been a review yet.

**Comment by pokgak** [comment]
I'm not really familiar with lwip but I compared this PR with how sock_udp_recv_buf is implemented and used in gnrc and overall it looks good to me. For testing, I ran the command described in the testing procedure above and all tests completed successfully. I'm not sure how to tests using fragmented packets. Do you have an extra application written for that @miri64?

Sorry, I cannot add much to the review but I can help tests it locally to confirm that it is working.

**Comment by miri64** [comment]
> If this can be reviewed now, just squash - there hasn't been a review yet.

Squashed

**Comment by miri64** [comment]
> I'm not sure how to tests using fragmented packets. Do you have an extra application written for that @miri64?

No, not really. However, `tests/gnrc_udp` has IPv6 fragmentation activated per default. You can just use that for sending UDP packets or pinging the lwIP node with payloads greater than 1432 (to exceed the Ethernet SDU).

**Comment by benpicco** [comment]
Can this be tested entirely on `native`?

**Comment by miri64** [comment]
> Can this be tested entirely on `native`?

Yepp, the automated version of `tests/lwip` even only runs on `native` ^^"

**Comment by benpicco** [comment]
> Also **please, please, please** test with (IP) fragmented packets (so larger than MTU) for `sock_udp` with `tests/lwip`. I did not have any time to test that right now, but we really should, as I removed a copying loop and replaced it just with a pointer assignment.

I reckon `make test` will not cover that, right?
'larger than MTU' here means > 1280 bytes and the console commands take payload as hex string. I'm afraid there is no way to produce such a packet with that test.

**Comment by miri64** [comment]
> I reckon make test will not cover that, right?
No.
> 'larger than MTU' here means > 1280 bytes and the console commands take payload as hex string. I'm afraid there is no way to produce such a packet with that test.
As I wrote: use `tests/gnrc_udp` for that. Give me minute, then I will provide a testing procedure for that...

**Comment by miri64** [comment]
I adapted `tests/lwip` to be able to receive such large packets:

```diff
diff --git a/tests/lwip/common.h b/tests/lwip/common.h
index 4cea2573df..6d0778a6f6 100644
--- a/tests/lwip/common.h
+++ b/tests/lwip/common.h
@@ -29,7 +29,7 @@ extern "C" {
  * @brief   Application configuration
  * @{
  */
-#define SOCK_INBUF_SIZE         (256)
+#define SOCK_INBUF_SIZE         (1500)
 #define SERVER_MSG_QUEUE_SIZE   (8)
 #define SERVER_BUFFER_SIZE      (64)
 /**
```

Then I started 3 terminals:

<details><summary><strong>Terminal 1 (<code>tshark</code> for sniffing)</strong></summary>

might also show other packets; the important part is the `IPv6 fragment` part

```
$ tshark -i tap0
Capturing on 'tap0'
    1 0.000000000 fe80::d827:1dff:fea8:6424 → fe80::e0bc:7dff:fecb:f550 IPv6 1510 IPv6 fragment (off=0 more=y ident=0xbf684be8 nxt=17)
    2 0.000158170 fe80::d827:1dff:fea8:6424 → fe80::e0bc:7dff:fecb:f550 UDP 122 1337 → 1337 Len=1500
```

</details>

<details><summary><strong>Terminal 2 (<code>tests/lwip</code> as UDP server)</strong></summary>

```
$ make -C tests/lwip all term
[…]
> ifconfig
ifconfig
ET_00:  inet6 fe80::e0bc:7dff:fecb:f550

> udp server start 1337
udp server start 1337
Success: started UDP server on port 1337
> Received UDP data from [fe80::d827:1dff:fea8:6424]:1337:
00000000  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000010  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000020  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000030  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000040  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000050  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000060  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000070  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000080  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000090  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000000A0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000000B0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000000C0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000000D0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000000E0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000000F0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000100  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000110  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000120  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000130  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000140  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000150  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000160  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000170  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000180  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000190  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000001A0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000001B0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000001C0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000001D0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000001E0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000001F0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000200  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000210  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000220  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000230  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000240  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000250  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000260  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000270  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000280  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000290  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000002A0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000002B0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000002C0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000002D0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000002E0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000002F0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000300  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000310  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000320  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000330  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000340  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000350  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000360  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000370  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000380  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000390  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000003A0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000003B0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000003C0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000003D0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000003E0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000003F0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000400  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000410  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000420  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000430  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000440  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000450  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000460  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000470  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000480  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000490  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000004A0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000004B0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000004C0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000004D0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000004E0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000004F0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000500  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000510  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000520  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000530  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000540  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000550  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000560  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000570  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000580  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
00000590  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000005A0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000005B0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000005C0  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01  01
000005D0  01  01  01  01  01  01  01  01  01  01  01  01
```

(1500 in hex is 0x5dc so we received the correct number of bytes)

</details>

<details><summary><strong>Terminal 3 (<code>tests/gnrc_udp</code> as UDP client)</strong></summary>

```
$ make -C tests/gnrc_udp all term
> udp send fe80::e0bc:7dff:fecb:f550 1337 1500                              
udp send fe80::e0bc:7dff:fecb:f550 1337 1500
Success: send 1500 byte to [fe80::e0bc:7dff:fecb:f550]:1337
```
</details>

**Pull request merged**